src/server.rs

```diff
@@ -909,4 +912,97 @@ impl CryptoService for CryptoServiceServer {
+        let priv_key = bitcoin::secp256k1::SecretKey::from_slice(&priv_key).map_err(|_err| {
+            tonic::Status::new(tonic::Code::InvalidArgument, "invalid priv_key".to_string())
+        })?;
+        let secp = Secp256k1::new();
+        let pub_key = bitcoin::secp256k1::PublicKey::from_secret_key(&secp, &priv_key);
+        let pub_key: [u8; SECP256K1_PUB_KEY_LENGTH] = pub_key.serialize();
+        let response = Secp256k1PrivKeyToPubKeyResponse {
+            pub_key: Some(ConsensusHex::encode_hex(&pub_key)),
+        };
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn secp256k1_sign(
+        &self,
+        request: tonic::Request<Secp256k1SignRequest>,
+    ) -> std::result::Result<tonic::Response<Secp256k1SignResponse>, tonic::Status> {
+        let Secp256k1SignRequest { message, priv_key } = request.into_inner();
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<Secp256k1SignRequest>("message"))?
+            .decode_tonic::<Secp256k1SignRequest, _>("message")?;
+        let digest = sha256::Hash::hash(&message).to_byte_array();
+        let message = bitcoin::secp256k1::Message::from_digest(digest);
+        let priv_key: [u8; SECP256K1_PRIV_KEY_LENGTH] = priv_key
+            .ok_or_else(|| missing_field::<Secp256k1SignRequest>("priv_key"))?
+            .decode_tonic::<Secp256k1SignRequest, _>("priv_key")?;
+        let priv_key = bitcoin::secp256k1::SecretKey::from_slice(&priv_key).map_err(|_err| {
+            tonic::Status::new(tonic::Code::InvalidArgument, "invalid priv_key".to_string())
+        })?;
+        let secp = Secp256k1::new();
+        let signature = secp.sign_ecdsa(&message, &priv_key);
+        let signature = signature.serialize_compact();
+        let response = Secp256k1SignResponse {
+            signature: Some(ConsensusHex::encode_hex(&signature)),
+        };
+        Ok(tonic::Response::new(response))
+    }
+
+    async fn secp256k1_verify(
+        &self,
+        request: tonic::Request<Secp256k1VerifyRequest>,
+    ) -> std::result::Result<tonic::Response<Secp256k1VerifyResponse>, tonic::Status> {
+        let Secp256k1VerifyRequest {
+            message,
+            signature,
+            pub_key,
+        } = request.into_inner();
+        let message: Vec<u8> = message
+            .ok_or_else(|| missing_field::<Secp256k1VerifyRequest>("message"))?
+            .decode_tonic::<Secp256k1VerifyRequest, _>("message")?;
+        let digest = sha256::Hash::hash(&message).to_byte_array();
+        let message = bitcoin::secp256k1::Message::from_digest(digest);
+        let signature: Vec<u8> = signature
+            .ok_or_else(|| missing_field::<Secp256k1VerifyRequest>("signature"))?
+            .decode_tonic::<Secp256k1VerifyRequest, _>("signature")?;
+        let signature_len = signature.len();
+        let signature: [u8; SECP256K1_SIGNATURE_COMPACT_SERIALIZATION_LENGTH] =
+            signature.try_into().map_err(|_err| {
```

Comment:
You can use `invalid_field_value::<Secp256k1VerifyRequest>("signature")` here

Reply:
In this case I think explicitly returning a `Status` is better, because the error tells you what is wrong -- signature length. Also in the `map_err` scope `signature` is moved, so it would be awkward to try to pass it to `invalid_field_value`.